### PR TITLE
Support fp4 fp8 non gated moe

### PR DESCRIPTION
**Pull request by TomerBN-Nvidia**
## Motivation

Add support for FP8 and NVFP4 quantization for the upcoming NVIDIA Nemotron v3 models.

## Modifications

Add support for FP8 and NVFP4 in non-gated MoE models

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 1]
```diff
@@ -743,6 +774,51 @@ def apply(
...
+                ],
+                ep_size=layer.moe_ep_size,
+                ep_rank=layer.moe_ep_rank,
+                tp_size=layer.moe_tp_size,
+                tp_rank=layer.moe_tp_rank,
+                tune_max_num_tokens=next_power_of_2(x.shape[0]),
+                activation_type=activation
+            )[0]
+
+            from sglang.srt.layers.moe.token_dispatcher import StandardCombineInput
```
[the remark this answers is not in the document]
Comment (TomerBN-Nvidia) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:838]: is is aligned with other parts of the code

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 2]
```diff
@@ -1527,7 +1610,7 @@ def _slice_scale(w):
             ("w2", layer.w2_weight_scale),
         ]:
             assert (
-                weight_scale.shape[2] % 16 == 0
+                weight_scale.shape[1] % 16 == 0
             ), f"Expected {name}_weight_scale.dim(2) to be divisible by 16"
```
[the remark this answers is not in the document]
Comment (TomerBN-Nvidia) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: refactored this part

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 3]
```diff
@@ -1601,6 +1684,7 @@ def _slice_scale(w):
             # Both flashinfer cutlass and regular cutlass use same processing for w2
 
             # Set up CUTLASS MoE parameters
+            num_shards = 2 if layer.moe_runner_config.is_gated else 1
```
[the remark this answers is not in the document]
Comment (TomerBN-Nvidia) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: done

**Comment by TomerBN-Nvidia** [comment]
@ispobock > @TomerBN-Nvidia Could you solve the conflict?

Done

File: python/sglang/srt/server_args.py [hunk 4]
```diff
@@ -1486,9 +1503,11 @@ def _handle_data_parallelism(self):
...
     def _handle_moe_kernel_config(self):
         if self.moe_runner_backend == "flashinfer_cutlass":
-            assert (
-                self.quantization == "modelopt_fp4" or self.quantization is None
-            ), "modelopt_fp4 quantization or bf16 is required for Flashinfer Cutlass MOE"
+            assert self.quantization in [
+                "modelopt_fp4",
+                "modelopt_fp8",
+                None,
+            ], "modelopt_fp4/8 quantization is required for Flashinfer Cutlass MOE"
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Qualified quantization methods should be `modelopt_fp4, modelopt_fp8 and bfloat16` here. This assertion message should be changed

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 from typing import TYPE_CHECKING, Any, Dict, List, Optional
 
 import torch
+from flashinfer.fused_moe.core import ActivationType
```
Comment (Fridge003) [inline python/sglang/srt/layers/quantization/modelopt_quant.py]: Should be moved to around line 90.  Needs try-catch for detecting ImportError

**Comment by Fridge003** [comment]
cc @ch-wan Can you have a look?

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 6]
```diff
@@ -1521,15 +1620,15 @@ def _slice_scale(w):
                 )
             }
         )
-
+        assert_dim = 2 if layer.moe_runner_config.is_gated else 1
```
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/modelopt_quant.py]: Move it below to the comment in the next line

**Review by ch-wan — APPROVED** [review]
LGTM. I only have one minor comment.

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 6 above)
Comment (TomerBN-Nvidia) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: Done

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 5 above)
Comment (TomerBN-Nvidia) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: Fixed
